### PR TITLE
Adding Custom SMTP Server Config

### PR DESCRIPTION
**Pull request by James9074**
## Description

I've added the following env vars which will tell nodemailer to use smtp auth instead of a [Well-known service](https://nodemailer.com/smtp/well-known/)

```BASH
MAILER_SMTP_HOST=
MAILER_SMTP_PORT=465
MAILER_SMTP_SECURE=true
```

To do this, I simply check if the `MAILER_SMTP_HOST` var is set, and if so, change the `options` property of the mailer config object to use the correct option set for connecting any SMTP server as documented in the [Nodemailer docs](https://nodemailer.com/smtp/):

```javascript
	mailer: {
		from: process.env.MAILER_FROM || 'no-reply@tellform.com',
		options: process.env.MAILER_SMTP_HOST ? { //Uses custom SMTP if MAILER_SMTP_HOST is set
			host: process.env.MAILER_SMTP_HOST || '',
			port: process.env.MAILER_SMTP_PORT || 465,
			secure: process.env.MAILER_SMTP_SECURE || true,
			auth: {
				user: process.env.MAILER_EMAIL_ID || '',
				pass: process.env.MAILER_PASSWORD || ''
			}
		} : {
			service: process.env.MAILER_SERVICE_PROVIDER || '',
			auth: {
				user: process.env.MAILER_EMAIL_ID || '',
				pass: process.env.MAILER_PASSWORD || ''
			}
		}
	}
```

> Don't worry, I changed the above in all environments and made sure to keep the same formatting and fallback values for each respective env.

I also added a check on server start in `user.server.model.js` right after the nodemailer transporter is created that verifies the connection and logs if it's bad. To do this, I had to update nodemailer from version 1.x to 4.x. I did all of my testing after that upgrade, and ensured that `nodemailer-sendgrid-transport` and `nodemailer-sparkpost-transport` would still work. Luckily for us, nodemailer hasn't done much to break it's API _post_ 1.0.

```javascript
var smtpTransport = nodemailer.createTransport(config.mailer.options);

// verify connection configuration on startup
smtpTransport.verify(function(error, success) {
  if (error) {
    console.log('Your mail configuration is incorrect', error);
  } else {
    console.log('Mail server is ready to take our messages');
  }
});
```

I also changed the node/npm version checks in package.json from node `6.11.1` to `6.11.2` and npm from `3.6.6` to `3.10.10`. This is because the Dockerfile pulls the latest 6.x version, and of course npm wasn't too happy to start a server with mismatched node/npm version requirements:

```docker
# Install nodejs
RUN curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
```

Finally, I removed a `debugger;` statement from the mailer which was causing my application to stop in it's tracks. I noticed it was the only place in the entire code base where `debugger;` appears, so I assume this isn't a critical logging feature.

## Motivation and Context

I'm deploying this in a restricted environment that can't use external mailing services (plus we just want to use the internal email server). Sadly, there's no way to effectively use this app on a multi-user level without email integration.

[Here you go](https://github.com/tellform/tellform/issues/220)

## How Has This Been Tested?

I've tested the following scenarios with both the Development and Production flags:

- Registration email alerts are sent with MailGun and Sendgrid using the same env file I used before making these changes
- Registration email alerts are sent with my new env variables using my own SMTP server if I provide a `MAILER_SMTP_HOST` env var
- Account creation works with both methods, verification emails are sent and I am able to create forms as usual with the newly created account
- When using incorrect credentials, either a bad API key _or_ a bad SMTP auth set, on server start I get a console warning

## Screenshots (if appropriate):
**The user will see this message if their service creds or smtp creds are invalid on server start:**
bad-creds

**The user will see this message on server start if the mail server accepted their creds:**
good-creds

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

I've only checked Breaking change since the `debugger;` line was causing the app to never send emails, even with MailGun/SendGrid

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

**Note** I didn't add unit tests as there were none for the current mailing system.

The goal is that you can pull this and verify it still works with your exact env config as it is now. I absolutely don't want to put something up that requires users to change their config.

### CODE REVIEW

File: app/models/user.server.model.js [hunk 1]
```diff
@@ -15,6 +15,14 @@ var mongoose = require('mongoose'),
 
 var smtpTransport = nodemailer.createTransport(config.mailer.options);
 
+// verify connection configuration on startup
+smtpTransport.verify(function(error, success) {
```
Comment (whitef0x0) [inline app/models/user.server.model.js]: This should probably been in https://github.com/tellform/tellform/blob/master/app/controllers/users/users.password.server.controller.js instead

File: config/env/development.js [hunk 2]
```diff
@@ -43,7 +43,15 @@ module.exports = {
 	},
 	mailer: {
 		from: process.env.MAILER_FROM || 'no-reply@tellform.com',
-		options: {
+		options: process.env.MAILER_SMTP_HOST ? { //Uses custom SMTP if MAILER_SMTP_HOST is set
```
Comment (whitef0x0) [inline config/env/development.js]: Maybe more clear to have this as an if statement instead?

Comment (James9074) [inline config/env/development.js]: I actually went with this due to the nature of how the configs are set up. To use a proper if statement I'd have to do this:

```javascript
mailer: {
  from: process.env.MAILER_FROM || 'no-reply@tellform.com',
  options: () => {
    if(process.env.MAILER_SMTP_HOST !== undefined) return { //Uses custom SMTP if MAILER_SMTP_HOST is set
      host: process.env.MAILER_SMTP_HOST || '',
      port: process.env.MAILER_SMTP_PORT || 465,
      secure: process.env.MAILER_SMTP_SECURE || true,
      auth: {
        user: process.env.MAILER_EMAIL_ID || '',
        pass: process.env.MAILER_PASSWORD || ''
      }
    } else return {
      service: process.env.MAILER_SERVICE_PROVIDER || '',
      auth: {
        user: process.env.MAILER_EMAIL_ID || '',
        pass: process.env.MAILER_PASSWORD || ''
      }
    }
  }
}
```

I'll change that if you'd prefer.

File: app/models/user.server.model.js (hunk 1 above)
Comment (whitef0x0) [inline app/models/user.server.model.js]: Awesome!

File: config/env/development.js (hunk 2 above)
Comment (James9074) [inline config/env/development.js]: **Edit:** If I change this I have to change all references to mailer.options like so: `config.mailer.options()`.

It would be preferable and more efficient to stick with the ternary expression is possible.

Comment (whitef0x0) [inline config/env/development.js]: I agree. Just leave it the way it originally was.
